# Numpy100 Exercise
[これ](https://github.com/rougier/numpy-100/blob/master/100_Numpy_exercises.ipynb)を和訳しながらやってく.
先駆者の方のは[こちら](https://github.com/osawat/numpy-100/blob/master/100_Numpy_exercises.ja.ipynb)

悩んだ・分からず答え見た奴は所感と考察を絶対述べる縛りで行きます.(大体わかってないはず)  
調べた場合はそのリンクを載せていく.

## 1. numpy パッケージをnpとして読み込む（難易度１）

In [16]:
import numpy as np

## 2. numpyのバージョンを確認する（難易度１）

In [4]:
# これは知ってた
print(np.__version__)

1.17.2


In [22]:
# これは知らなかった、答え合わせの時に初めて知った
np.show_config()

blas_mkl_info:
    libraries = ['blas', 'cblas', 'lapack', 'pthread', 'blas', 'cblas', 'lapack']
    library_dirs = ['/Users/Zi/anaconda3/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/Users/Zi/anaconda3/include']
blas_opt_info:
    libraries = ['blas', 'cblas', 'lapack', 'pthread', 'blas', 'cblas', 'lapack', 'blas', 'cblas', 'lapack']
    library_dirs = ['/Users/Zi/anaconda3/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/Users/Zi/anaconda3/include']
lapack_mkl_info:
    libraries = ['blas', 'cblas', 'lapack', 'pthread', 'blas', 'cblas', 'lapack']
    library_dirs = ['/Users/Zi/anaconda3/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/Users/Zi/anaconda3/include']
lapack_opt_info:
    libraries = ['blas', 'cblas', 'lapack', 'pthread', 'blas', 'cblas', 'lapack', 'blas', 'cblas', 'lapack']
    library_dirs = ['/Users/Zi/anaconda3/lib']
    define_macr

#### 所感・考察・調査
何かblasとかcblasとかって書いてあるものに関しては[ここ](http://azalea.s35.xrea.com/blas/whatIsBlas.html)に色々書いてあった.  
numpyの裏で動いている関数群のこと,それぞれpipかcondaかで速度の違いもある模様.参考は[こちら](https://orizuru.io/blog/machine-learning/anaconda_pip_difference/).

## 3. サイズ10の零ベクトルを作成する（難易度１）

In [14]:
# 所感：元のが"null vector"ってなっててnullのベクトル？？ってなってなかなか悩んだ普通に零ベクトルのことだった
a=np.zeros(10)

In [15]:
a

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])

## 4. 行列のメモリサイズを確認するには？（難易度１）

In [20]:
from sys import getsizeof
print(a.size*a.itemsize)
print(a.nbytes)
print(getsizeof(a))

80
80
176


#### 所感・考察・調査
`size` とかは割と使うので行けたが `itemsize` が思い出せなかった.  
ちなみにそれぞれの挙動は以下のよう.

In [17]:
print(a.size)
print(a.itemsize)

10
8


そもそも行列のメモリサイズが`size*itemsize`で表されることを認識していなかった.  
いつもは`nbytes`で確認していた.

`getsizeof`(176)と`nbytes`(80)の値が違うのは

- nbytes 格納されているデータのサイズのみを返す
- getsizeof 配列の次元,データ型などの情報も含むオブジェクト全体のサイズを返す

ためである.らしい.[参照元](https://living-sun.com/ja/python/727350-nbytes-and-getsizeof-return-different-values-python-numpy-memory.html)

## 5. コマンドラインから`numpy`の`add`ファンクションのドキュメントを得るには？（難易度１）

In [51]:
# なぜかマジックコマンドでできなかった…
%run python -c "import numpy; numpy.info(numpy.add)"

ERROR:root:File `'python.py'` not found.


In [52]:
!python -c "import numpy; numpy.info(numpy.add)"

add(x1, x2, /, out=None, *, where=True, casting='same_kind', order='K', dtype=None, subok=True[, signature, extobj])

Add arguments element-wise.

Parameters
----------
x1, x2 : array_like
    The arrays to be added. If ``x1.shape != x2.shape``, they must be broadcastable to a common shape (which becomes the shape of the output).
out : ndarray, None, or tuple of ndarray and None, optional
    A location into which the result is stored. If provided, it must have
    a shape that the inputs broadcast to. If not provided or `None`,
    a freshly-allocated array is returned. A tuple (possible only as a
    keyword argument) must have length equal to the number of outputs.
where : array_like, optional
    This condition is broadcast over the input. At locations where the
    condition is True, the `out` array will be set to the ufunc result.
    Elsewhere, the `out` array will retain its original value.
    Note that if an uninitialized `out` array is created via the default
    ``out=None`

#### 所感・考察・調査
マジックコマンドが動かなかったのが謎,IPython系使ってるなら以下で良さそう.  
(このexerciseの主旨からはズレるかもだが…)

`?np.add`

IPython系での`%`,`!`の違いに関しては[こちら](https://qiita.com/2-propanol/items/f6f63c3f20981df2c844)を参考.

## 06. 5番目の要素だけ1の大きさ10の零ベクトルを作る（難易度1）

In [5]:
a6 = np.zeros(10)
a6[4]=1

In [6]:
a6

array([0., 0., 0., 0., 1., 0., 0., 0., 0., 0.])

## 07. 10~49の範囲のベクトルを作成する（難易度1）

In [13]:
np.arange(10,49)

array([10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26,
       27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43,
       44, 45, 46, 47, 48])

## 08. ベクトルを並びを反対にする(最初の要素を最後の要素に)（難易度1)

In [21]:
base_vec08=np.arange(0,10)

In [22]:
# スライスのステップを-1刻みにしているだけ
base_vec08[::-1]

array([9, 8, 7, 6, 5, 4, 3, 2, 1, 0])

## 09. 0~8の値をもつ3*3の行列を作成する

In [32]:
base09=np.arange(0,9)

In [33]:
np.reshape(base09,(3,3))

array([[0, 1, 2],
       [3, 4, 5],
       [6, 7, 8]])

## 10. `[1,2,0,0,4,0]`から要素が0でないインデックスを見つける

In [1]:
base10=[1,2,0,0,4,0]

In [4]:
np.nonzero(base10)

(array([0, 1, 4]),)

## 11. 3*3の単位行列を作成する

In [4]:
np.eye(3)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

## 12. 乱数で 3x3x3 配列を生成する

In [7]:
np.random.random((3,3,3))

array([[[0.64698119, 0.79098664, 0.45057342],
        [0.74922779, 0.32366326, 0.79330371],
        [0.24070534, 0.15037871, 0.5730676 ]],

       [[0.51589321, 0.78605253, 0.67038458],
        [0.66549276, 0.29005795, 0.6538739 ],
        [0.67766727, 0.55551283, 0.87247062]],

       [[0.37060274, 0.70163803, 0.72155643],
        [0.61621483, 0.38096021, 0.89084974],
        [0.67266053, 0.85761725, 0.52165019]]])

#### 所感・考察・調査
`(要素数)*(行)*(列)`の対応になる.

In [9]:
print(np.random.random((4,2,3)))

[[[0.03166782 0.00359767 0.61373709]
  [0.5971344  0.01542775 0.84651715]]

 [[0.93710677 0.40950211 0.67042249]
  [0.39162913 0.67455406 0.16881662]]

 [[0.57743991 0.12219726 0.40213736]
  [0.94556456 0.62278508 0.31476066]]

 [[0.26041811 0.58775263 0.44853459]
  [0.76702448 0.61625877 0.1082101 ]]]


## 13. 乱数で 10x10 配列を生成して、その最小値と最大値を見つける

In [14]:
ans13=np.random.random((10,10))
print(ans13.max(),ans13.min())

0.9801364446356572 0.010385206456931528


#### 所感・考察・調査
最大・最小値を探す軸を指定することも出来る。[参考](https://deepage.net/features/numpy-min.html)

In [39]:
print(ans13)
print('----------')
print(ans13.min(axis=0))
print(ans13.min(axis=1))

[[0.05751401 0.72211795 0.98013644 0.66865119 0.57007621 0.51845256
  0.41144533 0.95666071 0.07729288 0.09228036]
 [0.67294601 0.24660254 0.17248138 0.020831   0.92324727 0.26807405
  0.90337815 0.68955872 0.87097369 0.36512301]
 [0.95784357 0.85269763 0.3375519  0.33761108 0.36377851 0.21640982
  0.52784014 0.45257702 0.15534156 0.9310942 ]
 [0.64928365 0.71259138 0.73483234 0.77218381 0.56934422 0.05110716
  0.95938524 0.17467257 0.16610172 0.53128872]
 [0.28732008 0.02663799 0.42668761 0.81825172 0.60492255 0.34113319
  0.89562393 0.72221351 0.05781961 0.01038521]
 [0.66233687 0.52702982 0.68568322 0.5984184  0.09571522 0.08273443
  0.46545543 0.55881863 0.65195544 0.80187392]
 [0.86574577 0.69699005 0.26254643 0.10924694 0.44093504 0.90926972
  0.70870228 0.08222082 0.16279547 0.83006848]
 [0.67281996 0.29405391 0.58430115 0.68666776 0.78668749 0.36976259
  0.51268017 0.14418759 0.22267337 0.83681941]
 [0.07339476 0.86399882 0.07641763 0.82938761 0.62296328 0.19039746
  0.35711493

## 14. 大きさ30の乱数のベクトルを生成して、その平均を求める

In [40]:
ans14=np.random.random(30)
ans14.mean()

0.48181880359763873

## 15. 周囲が 1 で内部が 0 であるような2次元配列生成する

In [47]:
ans15=np.ones((10,10))
ans15[1:-1,1:-1]=0
print(ans15)

[[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]]


## 16. 存在する行列の周りに0のパティングを引く

In [8]:
ans16=np.ones((4,4))

In [10]:
np.pad(ans16,2)

array([[0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 1., 1., 1., 1., 0., 0.],
       [0., 0., 1., 1., 1., 1., 0., 0.],
       [0., 0., 1., 1., 1., 1., 0., 0.],
       [0., 0., 1., 1., 1., 1., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0.]])

## 17. 以下の式の結果は?

In [12]:
0 * np.nan

nan

In [13]:
np.nan == np.nan

False

In [14]:
np.inf > np.nan

False

In [15]:
np.nan - np.nan

nan

In [16]:
0.3 == 3 * 0.1

False

## 18. 対角成分の直下に 1,2,3,4 の成分を持つ 5x5 行列を生成する

In [9]:
ans18=np.random.random((5,5))

In [22]:
ans18=np.diag(np.arange(4)+1,k=-1)

In [23]:
print(ans18)

[[0 0 0 0 0]
 [1 0 0 0 0]
 [0 2 0 0 0]
 [0 0 3 0 0]
 [0 0 0 4 0]]


#### 所感・考察・調査
`np.diag`によって対角要素の抽出・指定した対角要素を持った行列を生成できる。[参考](https://note.nkmk.me/python-numpy-diag-diagonal/)

## 19. 8x8 行列を生成して、市松模様で埋める

In [33]:
ans19 = np.zeros((8,8), dtype=int)

In [34]:
ans19[1::2,::2]=1
ans19[::2,1::2]=1

In [35]:
ans19

array([[0, 1, 0, 1, 0, 1, 0, 1],
       [1, 0, 1, 0, 1, 0, 1, 0],
       [0, 1, 0, 1, 0, 1, 0, 1],
       [1, 0, 1, 0, 1, 0, 1, 0],
       [0, 1, 0, 1, 0, 1, 0, 1],
       [1, 0, 1, 0, 1, 0, 1, 0],
       [0, 1, 0, 1, 0, 1, 0, 1],
       [1, 0, 1, 0, 1, 0, 1, 0]])

#### 所感・考察・調査
そもそも市松模様がわかってなかった.
もう少し模様の塊が大きいものは以下のよう
`np.repeat`でもうまいことできそうだけど題意にそわないしめんどいのでパス

In [79]:
ans19_a=np.zeros((8,8), dtype=int)

In [80]:
ans19_a[::4,::4]=1 #1*1, 1*5, 5*1, 5*5に対応
ans19_a[::4,1::4]=1 #1*2, 1*6, 5*2, 5*6に対応
ans19_a[1::4,::4]=1 #2*1, 2*5, 6*1, 6*5に対応
ans19_a[1::4,1::4]=1 #2*2, 2*6, 6*2, 6*6に対応

In [83]:
ans19_a[2::4,2::4]=1 #3*3, 3*7, 7*3, 7*7に対応
ans19_a[2::4,3::4]=1 #3*4, 3*8, 7*4, 7*8に対応
ans19_a[3::4,2::4]=1 #4*3, 4*7, 8*3, 8*7に対応
ans19_a[3::4,3::4]=1 #4*4, 4*8, 8*4, 8*8に対応

In [85]:
ans19_a #これ以上は関数作った方が圧倒的に楽そう

array([[1, 1, 0, 0, 1, 1, 0, 0],
       [1, 1, 0, 0, 1, 1, 0, 0],
       [0, 0, 1, 1, 0, 0, 1, 1],
       [0, 0, 1, 1, 0, 0, 1, 1],
       [1, 1, 0, 0, 1, 1, 0, 0],
       [1, 1, 0, 0, 1, 1, 0, 0],
       [0, 0, 1, 1, 0, 0, 1, 1],
       [0, 0, 1, 1, 0, 0, 1, 1]])

## 20. shape属性が (6,7,8) の配列のとき、100 番目の要素の添え字 (x,y,z)は？

In [95]:
base20=np.random.random((6,7,8))

In [102]:
np.unravel_index(100, np.shape(base20))

(1, 5, 4)

#### 所感・考察・調査
[参考](http://prpr.hatenablog.jp/entry/2016/06/03/numpy%E3%81%AEunravel_index%28%29%E3%81%AE%E6%8C%99%E5%8B%95%E3%81%AB%E3%81%A4%E3%81%84%E3%81%A6)